RecoParticleFlow/PFProducer/plugins/linkers/TrackAndGSFLinker.cc

```diff
@@ -12,49 +14,64 @@ class TrackAndGSFLinker : public BlockElementLinkerBase {
         useConvertedBrems_(conf.getParameter<bool>("useConvertedBrems")),
         debug_(conf.getUntrackedParameter<bool>("debug", false)) {}
 
-  double testLink(const reco::PFBlockElement*, const reco::PFBlockElement*) const override;
+  double testLink(size_t ielem1,
+                  size_t ielem2,
+                  reco::PFBlockElement::Type type1,
+                  reco::PFBlockElement::Type type2,
+                  const ElementListConst& elements,
+                  const PFTables& tables,
+                  const reco::PFMultiLinksIndex& multilinks) const override;
 
 private:
   bool useKDTree_, useConvertedBrems_, debug_;
 };
 
 DEFINE_EDM_PLUGIN(BlockElementLinkerFactory, TrackAndGSFLinker, "TrackAndGSFLinker");
 
-double TrackAndGSFLinker::testLink(const reco::PFBlockElement* elem1, const reco::PFBlockElement* elem2) const {
-  constexpr reco::PFBlockElement::TrackType T_FROM_GAMMACONV = reco::PFBlockElement::T_FROM_GAMMACONV;
+double TrackAndGSFLinker::testLink(size_t ielem1,
+                                   size_t ielem2,
+                                   reco::PFBlockElement::Type type1,
+                                   reco::PFBlockElement::Type type2,
+                                   const ElementListConst& elements,
+                                   const PFTables& tables,
+                                   const reco::PFMultiLinksIndex& multilinks) const {
   double dist = -1.0;
-  const reco::PFBlockElementGsfTrack* gsfelem(nullptr);
-  const reco::PFBlockElementTrack* tkelem(nullptr);
-  if (elem1->type() < elem2->type()) {
-    tkelem = static_cast<const reco::PFBlockElementTrack*>(elem1);
-    gsfelem = static_cast<const reco::PFBlockElementGsfTrack*>(elem2);
+
+  size_t itrack_elem = 0;
+  size_t igsf_elem = 0;
+  if (type1 < type2) {
+    itrack_elem = ielem1;
+    igsf_elem = ielem2;
   } else {
-    tkelem = static_cast<const reco::PFBlockElementTrack*>(elem2);
-    gsfelem = static_cast<const reco::PFBlockElementGsfTrack*>(elem1);
+    itrack_elem = ielem2;
+    igsf_elem = ielem1;
   }
+  size_t itrack = tables.element_to_track_[itrack_elem];
+  size_t igsf = tables.element_to_gsf_[igsf_elem];
+
+  const auto kf_nn = tables.track_table_vertex_.get<pf::track::KfTrackRefIsNonNull>(itrack);
+  const auto kf_key = tables.track_table_vertex_.get<pf::track::KfTrackRefKey>(itrack);
+  const auto kf_base_key = tables.track_table_vertex_.get<pf::track::KfTrackRefBaseKey>(itrack);
+
+  if (tables.gsf_table_.get<pf::track::KfPFRecTrackRefIsNonNull>(igsf)) {
+    //const reco::TrackRef& gsftrackref = refkf->trackRef();
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

Reply:
yes, all these commented out pieces are legacy and should be removed - thanks for noticing.